src/brainfuck/virtualmachine/Machine.java

```diff
@@ -92,9 +94,13 @@ public byte readMemory() {
 	 * @param instr	Instruction to execute.
 	 */
 	public void executeOp(Instruction instr) {
-		if (!jumping) instr.accept(this); // Jumping may be modified in there
-		if (jumping && instr instanceof ConditionalJump) { // Increase the corresponding counter when encountering a conditional jump instruction
-			((ConditionalJump) instr).incr(bracketCounter);
+		if (instr instanceof Jump && readMemory() == Byte.MIN_VALUE) {
+			this.jumping = true;
+		} else if (instr instanceof Back && readMemory() != Byte.MIN_VALUE) {
+			this.jumping = true;
+		} else {
+			this.jumping = false ;
+			instr.accept(this);
```

Comment:
Not quite convinced by the memory value checks here, especially when the behaviour is the same for both instructions. It should be a method of ConditionalJump (replacing the already-existing one), exploiting polymorphism and dynamic typing.